src/video/psp/SDL_pspvideo.c

```diff
@@ -29,14 +29,119 @@
+    PSP_RenderData *rd;
+    int i, w, h;
+
+    SDL_GetWindowSizeInPixels(window, &w, &h);
+
+    if(w != 480)
+        return SDL_SetError("Unexpected window size");
+
+    if (!data) {
+        SDL_Renderer *renderer = NULL;
+        for (i = 0; i < SDL_GetNumRenderDrivers(); ++i) {
+            SDL_GetRenderDriverInfo(i, &info);
+            if (SDL_strcmp(info.name, "software") != 0) {
+                renderer = SDL_CreateRenderer(window, i, 0);
+                if (renderer && (SDL_GetRendererInfo(renderer, &info) == 0) && (info.flags & SDL_RENDERER_ACCELERATED)) {
+                    break; /* this will work. */
+                }
+                if (renderer) { /* wasn't accelerated, etc, skip it. */
+                    SDL_DestroyRenderer(renderer);
+                    renderer = NULL;
+                }
+            }
+        }
+        if (!renderer) {
+            return SDL_SetError("No hardware accelerated renderers available");
+        }
+
+        SDL_assert(renderer != NULL); /* should have explicitly checked this above. */
+
+        /* Create the data after we successfully create the renderer (bug #1116) */
+        data = (SDL_WindowTextureData *)SDL_calloc(1, sizeof(*data));
+
+        if (!data) {
+            SDL_DestroyRenderer(renderer);
+            return SDL_OutOfMemory();
+        }
+
+        SDL_SetWindowData(window, SDL_WINDOWTEXTUREDATA, data);
+        data->renderer = renderer;
+    } else {
+        if (SDL_GetRendererInfo(data->renderer, &info) == -1) {
+            return -1;
+        }
+    }
+
+    /* Find the first format without an alpha channel */
+    *format = info.texture_formats[0];
+
+    for (i = 0; i < (int)info.num_texture_formats; ++i) {
+        if (!SDL_ISPIXELFORMAT_FOURCC(info.texture_formats[i]) &&
+            !SDL_ISPIXELFORMAT_ALPHA(info.texture_formats[i])) {
+            *format = info.texture_formats[i];
+            break;
+        }
+    }
+
+    /* get the PSP renderer's "private" data */
+    rd = data->renderer->driverdata;
+
+    /* Create framebuffer data */
+    data->bytes_per_pixel = SDL_BYTESPERPIXEL(*format);
+    /* since we point directly to VRAM's frontbuffer, we have to use
+       the upscaled pitch of 512 width - since PSP requires all textures to be
+       powers of 2. */
+    data->pitch = 512 * data->bytes_per_pixel;
+    data->pixels = rd->frontbuffer;
+    *pixels = data->pixels;
+    *pitch = data->pitch;
+
+    /* Make sure we're not double-scaling the viewport */
+    SDL_RenderSetViewport(data->renderer, NULL);
+
+    return 0;
+}
+
+int PSP_UpdateWindowFramebuffer(_THIS, SDL_Window *window, const SDL_Rect *rects, int numrects) {
+    PSP_RenderData *rd;
+    SDL_WindowTextureData *data = SDL_GetWindowData(window, SDL_WINDOWTEXTUREDATA);
+    if(!data || !data->renderer || !window->surface) return -1;
+    rd = data->renderer->driverdata;
+    PSP_RenderPresent(data->renderer);
```

Comment:
Can not be done in a different way? to expose `PSP_RenderPresent` looks a code smell, is not possible to do something as:
```c
data->renderer->RenderPresent(data->renderer)
```

Reply:
good catch!